source/common/ssl/context_config_impl.cc

```diff
@@ -45,32 +47,57 @@ Secret::TlsCertificateConfigProviderSharedPtr getTlsCertificateConfigProvider(
 }
 
+Secret::CertificateValidationContextConfigProviderSharedPtr
+getCertificateValidationContextConfigProviderFromSds(
+    const envoy::api::v2::auth::SdsSecretConfig& sds_secret_config,
+    const envoy::api::v2::auth::Secret& default_secret) {
+  if (!sds_secret_config.has_sds_config()) {
+    // static secret
+    auto secret_provider =
+        factory_context.secretManager().findStaticCertificateValidationContextProvider(
+            sds_secret_config.name());
+    if (!secret_provider) {
+      throw EnvoyException(fmt::format("Unknown static certificate validation context: {}",
+                                       sds_secret_config.name()));
+    }
+    return secret_provider;
+  } else {
+    auto secret_provider =
+        factory_context.secretManager().findOrCreateCertificateValidationContextProvider(
+            sds_secret_config.sds_config(), sds_secret_config.name(), factory_context);
+    secret_provider->setDefaultSecret(default_cvc);
+    return secret_provider;
+  }
+  return nullptr;
+}
+
 Secret::CertificateValidationContextConfigProviderSharedPtr
 getCertificateValidationContextConfigProvider(
     const envoy::api::v2::auth::CommonTlsContext& config,
     Server::Configuration::TransportSocketFactoryContext& factory_context) {
-  if (config.has_validation_context()) {
+  switch (config.validation_context_type_case()) {
+  case envoy::api::v2::auth::CommonTlsContext::ValidationContextTypeCase::kValidationContext: {
     return factory_context.secretManager().createInlineCertificateValidationContextProvider(
         config.validation_context());
   }
-  if (config.has_validation_context_sds_secret_config()) {
+  case envoy::api::v2::auth::CommonTlsContext::ValidationContextTypeCase::
+      kValidationContextSdsSecretConfig: {
     const auto& sds_secret_config = config.validation_context_sds_secret_config();
-    if (!sds_secret_config.has_sds_config()) {
-      // static secret
-      auto secret_provider =
-          factory_context.secretManager().findStaticCertificateValidationContextProvider(
-              sds_secret_config.name());
-      if (!secret_provider) {
-        throw EnvoyException(fmt::format("Unknown static certificate validation context: {}",
-                                         sds_secret_config.name()));
-      }
-      return secret_provider;
-    } else {
-      return factory_context.secretManager().findOrCreateCertificateValidationContextProvider(
-          sds_secret_config.sds_config(), sds_secret_config.name(), factory_context);
-    }
+    envoy::api::v2::auth::Secret default_secret;
+    return getCertificateValidationContextConfigProviderFromSds(sds_secret_config, default_secret);
+  }
+  case envoy::api::v2::auth::CommonTlsContext::ValidationContextTypeCase::
+      kCombinedValidationContext: {
+    envoy::api::v2::auth::Secret default_secret;
+    default_secret.mutable_validation_context()->CopyFrom(
+        config.combined_validation_context().default_validation_context());
```

Comment:
Not need to make a copy. Just do

const auto& default_secret = config.combined_validation_context().default_validation_context();

BTW,  setDefaultSecret() not need to be at  base class.  It can be specific to CertificateValidationContextProvider

Reply:
Done. Thanks.